skops/io/_audit.py

```diff
@@ -59,3 +72,284 @@ def audit_tree(tree, trusted):
+        or are trusted by default.
+
+    memoize : bool, default=True
+        If ``True``, the object will be memoized in the load context, if it has
+        the ``__id__`` set. This is used to avoid loading the same object
+        multiple times.
+    """
+
+    def __init__(
+        self,
+        state: dict[str, Any],
+        load_context: LoadContext,
+        trusted: bool | Sequence[str] = False,
+        memoize: bool = True,
+    ) -> None:
+        self.class_name, self.module_name = state["__class__"], state["__module__"]
+        self._is_safe = None
+        self._constructed = UNINITIALIZED
+        saved_id = state.get("__id__")
+        if saved_id and memoize:
+            # hold reference to obj in case same instance encountered again in
+            # save state
+            load_context.memoize(self, saved_id)
+
+        # subclasses should always:
+        # 1. call super().__init__()
+        # 2. set self.trusted = self._get_trusted(trusted, ...) where ... is a
+        #    list of appropriate trusted types
+        # 3. set self.children, where children are states of child nodes; do not
+        #    construct the children objects yet
+        self.trusted = self._get_trusted(trusted, [])
+        self.children: dict[str, Any] = {}
+
+    def construct(self) -> Any:
+        """Construct the object.
+
+        We only construct the object once, and then cache the result.
+        """
+        if self._constructed is not UNINITIALIZED:
+            return self._constructed
+        self._constructed = self._construct()
+        return self._constructed
+
+    def _construct(self) -> Any:
+        raise NotImplementedError(
+            f"{self.__class__.__name__} should implement a 'construct' method"
+        )
+
+    @staticmethod
+    def _get_trusted(
+        trusted: bool | Sequence[str], default: list[str]
+    ) -> Literal[True] | list[str]:
+        """Return a trusted list, or True.
+
+        If ``trusted`` is ``False``, we return the ``default``, otherwise the
+        ``trusted`` value is used.
+
+        This is a convenience method called by child classes.
+        """
+        if trusted is True:
+            # if trusted is True, we trust the node
+            return True
+
+        if trusted is False:
+            # if trusted is False, we only trust the defaults
+            return default
+
+        # otherwise we trust the given list, call list in case it's a tuple
+        return list(trusted)
+
+    def is_self_safe(self) -> bool:
+        """True only if the node's type is considered safe.
+
+        This property only checks the type of the node, not its children.
+        """
+        return check_type(self.module_name, self.class_name, self.trusted)
+
+    def is_safe(self) -> bool:
+        """True only if the node and all its children are safe."""
+        # if trusted is set to True, we don't do any safety checks.
+        if self.trusted is True:
+            return True
```

Comment:
I know you just copied this over, but we should probably should add a test that covers this

Reply:
I discussed this with Adrin, the code is currently not being used. For context, there was once the idea of adding a feature based on it, but that didn't manifest, so we'll remove it in a separate PR.